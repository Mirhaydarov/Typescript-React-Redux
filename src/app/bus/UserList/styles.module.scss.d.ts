export const wrap: string;
export const userList: string;
