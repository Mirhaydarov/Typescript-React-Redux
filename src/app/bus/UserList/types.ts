export type UserList = {
  id: number;
  name: string;
  email: string;
  phone: string;
  image: string;
  address: {
    street: string;
    city: string;
  };
  company: {
    name: string;
    catchPhrase: string;
  };
};
export type UserPhoto = {
  userImageURL: string;
};

export type UsersList = UserList[];
export type UsersPhoto = { hits: UserPhoto[] };
export type HttpErrorType = { stack: string };

export const USER_LIST_START_FETCHING = 'USER_LIST_START_FETCHING';
type UserListStartFetchingAction = {
  type: typeof USER_LIST_START_FETCHING;
};

export const USER_LIST_STOP_FETCHING = 'USER_LIST_STOP_FETCHING';
type UserListStopFetchingAction = {
  type: typeof USER_LIST_STOP_FETCHING;
};

export const USER_LIST_SUCCESS_FETCHING =
  'USER_LIST_SUCCESS_FETCHING';
export type UserListSuccessFetchingAction = {
  type: typeof USER_LIST_SUCCESS_FETCHING;
  payload: UsersList;
};

export const USER_LIST_ERROR_FETCHING = 'USER_LIST_ERROR_FETCHING';
export type UserListErrorFetchingAction = {
  type: typeof USER_LIST_ERROR_FETCHING;
  error: boolean;
  payload: HttpErrorType;
};

export const USER_LIST_ASYNC_FETCHING = 'USER_LIST_ASYNC_FETCHING';
type UserListAsyncFetchingAction = {
  type: typeof USER_LIST_ASYNC_FETCHING;
};

export type UserListActionTypes =
  | UserListStartFetchingAction
  | UserListStopFetchingAction
  | UserListErrorFetchingAction
  | UserListSuccessFetchingAction
  | UserListAsyncFetchingAction;
