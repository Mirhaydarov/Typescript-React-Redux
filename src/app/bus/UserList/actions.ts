import {
  USER_LIST_START_FETCHING,
  USER_LIST_STOP_FETCHING,
  USER_LIST_SUCCESS_FETCHING,
  USER_LIST_ERROR_FETCHING,
  USER_LIST_ASYNC_FETCHING,
  UserListActionTypes,
  UserListSuccessFetchingAction,
  UserListErrorFetchingAction,
  UsersList,
  HttpErrorType,
} from './types';

export function startFetching(): UserListActionTypes {
  return {
    type: USER_LIST_START_FETCHING,
  };
}

export function stopFetching(): UserListActionTypes {
  return {
    type: USER_LIST_STOP_FETCHING,
  };
}

export function successFetching(
  payload: UsersList
): UserListSuccessFetchingAction {
  return {
    type: USER_LIST_SUCCESS_FETCHING,
    payload,
  };
}

export function errorFetching(
  payload: HttpErrorType
): UserListErrorFetchingAction {
  return {
    type: USER_LIST_ERROR_FETCHING,
    error: false,
    payload,
  };
}

export function asyncFetching(): UserListActionTypes {
  return {
    type: USER_LIST_ASYNC_FETCHING,
  };
}
