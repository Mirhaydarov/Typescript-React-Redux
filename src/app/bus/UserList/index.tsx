// Core
import React, { FC, ReactElement } from 'react';

// Components
import { Spinner } from '../../components/Spinner';
import { ErrorIndicator } from '../../components/ErrorIndicator';
import { UserListTile } from '../../components/UserListTile';

// Hooks
import { useUserListFetch } from './hooks/useUserListFetch';

// Styles
import styles from './styles.module.scss';

export const UserList: FC = (): ReactElement => {
  const { isFetching, data, error } = useUserListFetch();

  const ErrorMessageJSX = error && <ErrorIndicator />;
  const LoaderJSX = isFetching && <Spinner />;
  const UserListJSX = isFetching || <UserListTile data={data} />;

  return (
    <section className={styles.userList}>
      <div className={styles.wrap}>
        {LoaderJSX}
        {ErrorMessageJSX}
        {UserListJSX}
      </div>
    </section>
  );
};
