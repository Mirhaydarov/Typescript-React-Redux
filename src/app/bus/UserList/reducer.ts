import {
  USER_LIST_START_FETCHING,
  USER_LIST_STOP_FETCHING,
  USER_LIST_SUCCESS_FETCHING,
  USER_LIST_ERROR_FETCHING,
  USER_LIST_ASYNC_FETCHING,
  UserListActionTypes,
  UsersList,
  HttpErrorType,
} from './types';

export type UserListState = {
  data: UsersList;
  isFetching: boolean;
  error: boolean | HttpErrorType;
};

const initialState: UserListState = {
  data: [],
  isFetching: false,
  error: true,
};

export function userListReducer(
  state: UserListState = initialState,
  action: UserListActionTypes
): UserListState {
  switch (action.type) {
    case USER_LIST_START_FETCHING:
      return {
        ...state,
        isFetching: true,
        error: false,
      };

    case USER_LIST_STOP_FETCHING:
      return {
        ...state,
        isFetching: false,
      };

    case USER_LIST_SUCCESS_FETCHING:
      return {
        ...state,
        data: action.payload,
        error: false,
      };

    case USER_LIST_ERROR_FETCHING:
      return {
        ...state,
        error: action.payload,
      };

    case USER_LIST_ASYNC_FETCHING:
      return state;

    default:
      // eslint-disable-next-line no-case-declarations, @typescript-eslint/no-unused-vars
      const never: never = action;
  }
  return state;
}
