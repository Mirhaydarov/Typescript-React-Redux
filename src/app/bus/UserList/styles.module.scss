.wrap {
  composes: wrap from '../../../root-styles.scss';
}

.user-list {
  padding-top: 50px;
  padding-bottom: 50px;
}