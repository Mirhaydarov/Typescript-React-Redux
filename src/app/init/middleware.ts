// Core
import { Middleware } from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import { createBrowserHistory } from 'history';
import { routerMiddleware as createRouterMiddleware } from 'connected-react-router';

const logger = createLogger({
  duration: true,
  collapsed: true,
  diff: true,
  colors: {
    title: (action): string => (action.error ? '#ff0005' : '#1C5FAF'),
    prevState: (): string => '#1C5FAF',
    action: (): string => '#149945',
    nextState: (): string => '#fcff40',
    error: (): string => '#ff0005',
  },
});

const sagaMiddleware = createSagaMiddleware();
const history = createBrowserHistory();
const routerMiddleware = createRouterMiddleware(history);

const middleware: Middleware[] = [sagaMiddleware, routerMiddleware];

const devEnvironment = process.env.NODE_ENV === 'development';

if (devEnvironment) middleware.push(logger);

export { history, middleware, sagaMiddleware };
