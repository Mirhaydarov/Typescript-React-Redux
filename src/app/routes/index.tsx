// Core
import React, { FC, ReactElement } from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';

// Route path
import { book } from './book';

export const pageNotFound = (): ReactElement => (
  <h2>404: Not Found</h2>
);

export const Routes: FC = (): ReactElement => (
  <>
    <Switch>
      <Route component={pageNotFound} />
      <Redirect to={book.root} />
    </Switch>
  </>
);