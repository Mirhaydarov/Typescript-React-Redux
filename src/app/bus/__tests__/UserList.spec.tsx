// Core
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import fetchMock from 'fetch-mock';

// Components
import { UserList } from '../UserList';

// Instruments
import {
  makeMountRender,
  reduxify,
  mocker,
  store,
  loaderStore,
  errorStore,
  initialState,
} from '../../../test-utils';

// Api
import { api } from '../../../api';
import {
  root,
  rootPhoto,
  photoApiKey,
  lengthPhoto,
} from '../../../api/config';

// Actions
import {
  asyncFetching,
  startFetching,
  stopFetching,
  errorFetching,
  successFetching,
} from '../UserList/actions';
import { history } from '../../init/middleware';

// Types
import {
  USER_LIST_START_FETCHING,
  USER_LIST_STOP_FETCHING,
  USER_LIST_ERROR_FETCHING,
  USER_LIST_SUCCESS_FETCHING,
  USER_LIST_ASYNC_FETCHING,
} from '../UserList/types';

// Routes
import { Routes, pageNotFound } from '../../routes';

describe('<UserList />', () => {
  beforeEach(() => {
    mocker(fetchMock).fetchUsers();
    mocker(fetchMock).fetchUsersPhoto();

    store.clearActions();
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('Dispatches the correct action and payload', () => {
    const expectedAsyncActions = { type: USER_LIST_ASYNC_FETCHING };
    const expectedStartActions = { type: USER_LIST_START_FETCHING };
    const expectedErrorActions = {
      type: USER_LIST_ERROR_FETCHING,
      error: false,
      payload: { stack: '' },
    };
    const expectedSuccessActions = {
      type: USER_LIST_SUCCESS_FETCHING,
      payload: [...initialState.userList.data],
    };
    const expectedStopActions = { type: USER_LIST_STOP_FETCHING };

    store.dispatch(asyncFetching());
    store.dispatch(startFetching());
    store.dispatch(errorFetching({ stack: '' }));
    store.dispatch(successFetching([...initialState.userList.data]));
    store.dispatch(stopFetching());

    expect(store.getActions()).toEqual([
      expectedAsyncActions,
      expectedStartActions,
      expectedErrorActions,
      expectedSuccessActions,
      expectedStopActions,
    ]);
  });

  it('Fetching is correct', () => {
    api.getUsers.fetch();
    api.getUsersPhoto.fetch();

    expect(fetchMock.called(`${root}/users`)).toBe(true);
    expect(
      fetchMock.called(
        `${rootPhoto}/?key=${photoApiKey}&image_type=photo&per_page=${lengthPhoto}&safesearch=true`
      )
    ).toBe(true);
  });

  it('Matches snapshot', () => {
    const wrapper = makeMountRender(reduxify(Routes, store));

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Allows users to select a profile', () => {
    const wrapper = makeMountRender(reduxify(Routes, store));

    history.push('/profile/leanne-graham');
    wrapper.update();

    wrapper.find('.tableBody > tr').simulate('click');

    wrapper.update();
    expect(window.location.href).toBe(
      'http://localhost/profile/leanne-graham'
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should render a <Spinner />', () => {
    const wrapper = makeMountRender(reduxify(UserList, loaderStore));

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should render an <ErrorIndicator />', () => {
    const wrapper = makeMountRender(reduxify(UserList, errorStore));

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should render a page not found if url to incorrect', (done) => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/wrong-path']}>
          <Routes />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find(pageNotFound)).toHaveLength(1);

    history.push('/wrong-path');
    wrapper.update();

    process.nextTick(() => {
      wrapper.update();
      expect(window.location.href).toBe(
        'http://localhost/wrong-path'
      );
      done();
    });
  });
});
