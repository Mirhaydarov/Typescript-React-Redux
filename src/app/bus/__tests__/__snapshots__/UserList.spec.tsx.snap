// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> Allows users to select a profile 1`] = `
<header class="header">
  <nav class="nav">
    <a class="navLink"
       href="/"
    >
      Home
    </a>
    <input class="searchBar"
           type="search"
           placeholder="Search people by name..."
           value
    >
  </nav>
</header>
<section class="profile">
  <div class="wrap">
    <h2 class="title">
      Profile:&nbsp;Leanne Graham
    </h2>
    <div class="listContainer">
      <img class="avatar"
           src="https://cdn.pixabay.com/user/2020/10/03/07-51-18-705_250x250.jpg"
           alt="Leanne Graham"
      >
      <div>
        <ul class="listUser">
          <li class="listItem">
            <span class="listItemDesc">
              Name:
            </span>
            &nbsp;Leanne Graham
          </li>
          <li class="listItem">
            <span class="listItemDesc">
              Email:
            </span>
            &nbsp;Sincere@april.biz
          </li>
          <li class="listItem">
            <span class="listItemDesc">
              Phone:
            </span>
            &nbsp;1-770-736-8031 x56442
          </li>
        </ul>
      </div>
      <div>
        <h3 class="listTitle">
          Company
        </h3>
        <ul class="listCompany">
          <li class="listItem">
            <span class="listItemDesc">
              Name:
            </span>
            &nbsp;Romaguera-Crona
          </li>
          <li class="listItem">
            <span class="listItemDesc">
              Catch phrase:
            </span>
            &nbsp;Multi-layered client-server neural-net
          </li>
        </ul>
      </div>
      <div>
        <h3 class="listTitle">
          Address
        </h3>
        <ul class="listAddress">
          <li class="listItem">
            <span class="listItemDesc">
              City:
            </span>
            &nbsp;Gwenborough
          </li>
          <li class="listItem">
            <span class="listItemDesc">
              Street:
            </span>
            &nbsp;Kulas Light
          </li>
        </ul>
      </div>
    </div>
  </div>
</section>
`;

exports[`<UserList /> Matches snapshot 1`] = `
<header class="header">
  <nav class="nav">
    <input class="searchBar"
           type="search"
           placeholder="Search people by name..."
           value
    >
  </nav>
</header>
<section class="userList">
  <div class="wrap">
    <table class="table">
      <thead class="tableHead">
        <tr>
          <th>
            Image
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Email
          </th>
        </tr>
      </thead>
      <tbody class="tableBody">
        <tr>
          <td>
            <img class="tableBodyImage"
                 src="https://cdn.pixabay.com/user/2020/10/03/07-51-18-705_250x250.jpg"
                 alt="Leanne Graham"
            >
          </td>
          <td>
            Leanne Graham
          </td>
          <td>
            <a class="tableBodyPhone"
               href="tel:+1-770-736-8031 x56442"
            >
              1-770-736-8031 x56442
            </a>
          </td>
          <td>
            <a class="tableBodyEmail"
               href="mailto:Sincere@april.biz"
            >
              Sincere@april.biz
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`<UserList /> Should render a <Spinner /> 1`] = `
<section class="userList">
  <div class="wrap">
    <div class="spinner">
      <div aria-busy="true"
           class
      >
        <svg height="80"
             width="80"
             fill="#89a8ff"
             viewbox="0 0 55 80"
             xmlns="http://www.w3.org/2000/svg"
             aria-label="audio-loading"
        >
          <g transform="matrix(1 0 0 -1 0 80)">
            <rect width="10"
                  height="20"
                  rx="3"
            >
              <animate attributename="height"
                       begin="0s"
                       dur="4.3s"
                       values="20;45;57;80;64;32;66;45;64;23;66;13;64;56;34;34;2;23;76;79;20"
                       calcmode="linear"
                       repeatcount="indefinite"
              >
              </animate>
            </rect>
            <rect x="15"
                  width="10"
                  height="80"
                  rx="3"
            >
              <animate attributename="height"
                       begin="0s"
                       dur="2s"
                       values="80;55;33;5;75;23;73;33;12;14;60;80"
                       calcmode="linear"
                       repeatcount="indefinite"
              >
              </animate>
            </rect>
            <rect x="30"
                  width="10"
                  height="50"
                  rx="3"
            >
              <animate attributename="height"
                       begin="0s"
                       dur="1.4s"
                       values="50;34;78;23;56;23;34;76;80;54;21;50"
                       calcmode="linear"
                       repeatcount="indefinite"
              >
              </animate>
            </rect>
            <rect x="45"
                  width="10"
                  height="30"
                  rx="3"
            >
              <animate attributename="height"
                       begin="0s"
                       dur="2s"
                       values="30;45;13;80;56;72;45;76;34;23;67;30"
                       calcmode="linear"
                       repeatcount="indefinite"
              >
              </animate>
            </rect>
          </g>
        </svg>
      </div>
    </div>
  </div>
</section>
`;

exports[`<UserList /> Should render an <ErrorIndicator /> 1`] = `
<section class="userList">
  <div class="wrap">
    <div class="errorIndicator">
      <h2>
        Error
      </h2>
      <p>
        Something has gone terribly wrong
      </p>
      <p>
        (but we already try to fix it)
      </p>
    </div>
    <table class="table">
      <thead class="tableHead">
        <tr>
          <th>
            Image
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Email
          </th>
        </tr>
      </thead>
      <tbody class="tableBody">
        <tr>
          <td>
            <img class="tableBodyImage"
                 src="https://cdn.pixabay.com/user/2020/10/03/07-51-18-705_250x250.jpg"
                 alt="Leanne Graham"
            >
          </td>
          <td>
            Leanne Graham
          </td>
          <td>
            <a class="tableBodyPhone"
               href="tel:+1-770-736-8031 x56442"
            >
              1-770-736-8031 x56442
            </a>
          </td>
          <td>
            <a class="tableBodyEmail"
               href="mailto:Sincere@april.biz"
            >
              Sincere@april.biz
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
